File: pythonlibrary/aibrd/analyzers/test_linkage.py
"""Test linkage analyzer — maps requirement IDs to actual test files in the workspace."""

from __future__ import annotations
import re
import subprocess
from dataclasses import dataclass, field
from datetime import date
from pathlib import Path


TEST_PATTERNS = re.compile(
    r"\.(test|spec)\.(ts|tsx|js|jsx|py|java)$|Test\.(java|py)$|_test\.(py|go)$"
)


@dataclass
class TestLink:
    requirement_id: str
    requirement_title: str
    test_files: list[str] = field(default_factory=list)
    test_count: int = 0
    covered: bool = False


@dataclass
class TestLinkageReport:
    covered_count: int
    uncovered_count: int
    total_tests: int
    links: list[TestLink] = field(default_factory=list)


def _git_ls_files(workspace_root: str) -> list[str]:
    try:
        out = subprocess.check_output(
            ["git", "-C", workspace_root, "ls-files"],
            encoding="utf-8", stderr=subprocess.DEVNULL,
        )
        return [f for f in out.splitlines() if TEST_PATTERNS.search(f)]
    except Exception:
        return []


def link_test_files(
    aibrd_dir: str,
    workspace_root: str,
    slug: str | None = None,
) -> list[TestLink]:
    ctx_path = (
        Path(aibrd_dir) / "modules" / slug / "CONTEXT.md"
        if slug
        else Path(aibrd_dir) / "CONTEXT.md"
    )
    if not ctx_path.exists():
        return []

    text = ctx_path.read_text(encoding="utf-8")
    test_files = _git_ls_files(workspace_root)
    ws = Path(workspace_root)
    links: list[TestLink] = []

    for m in re.finditer(r"###\s+((?:[A-Z]+-)?(?:BF|AC|TC)-\d+):\s*(.+)", text):
        req_id, title = m.group(1), m.group(2).strip()
        matched: list[str] = []
        for tf in test_files:
            try:
                src = (ws / tf).read_text(encoding="utf-8", errors="ignore")
                if req_id in src:
                    matched.append(tf)
            except Exception:
                pass
        links.append(TestLink(
            requirement_id=req_id,
            requirement_title=title,
            test_files=matched,
            test_count=len(matched),
            covered=len(matched) > 0,
        ))

    return links


def build_test_linkage_report(links: list[TestLink]) -> TestLinkageReport:
    return TestLinkageReport(
        covered_count=sum(1 for l in links if l.covered),
        uncovered_count=sum(1 for l in links if not l.covered),
        total_tests=sum(l.test_count for l in links),
        links=links,
    )


def format_test_linkage_report(report: TestLinkageReport, slug: str | None = None) -> str:
    today = date.today().isoformat()
    scope = f" — {slug}" if slug else ""
    total = len(report.links)
    pct = round(report.covered_count / total * 100) if total else 0

    lines = [
        f"# Test Linkage Report{scope}",
        f"_Generated: {today}_",
        "",
        f"> **{pct}% requirements covered** · {report.covered_count} covered · {report.uncovered_count} uncovered · {report.total_tests} test file references",
        "",
    ]

    uncovered = [l for l in report.links if not l.covered]
    covered = [l for l in report.links if l.covered]

    if uncovered:
        lines += ["## ❌ Not Covered by Tests", ""]
        lines.append("| Requirement | Title |")
        lines.append("|---|---|")
        for l in uncovered:
            lines.append(f"| **{l.requirement_id}** | {l.requirement_title[:60]} |")
        lines.append("")

    if covered:
        lines += ["## ✅ Covered", ""]
        lines.append("| Requirement | Title | Test Files |")
        lines.append("|---|---|---|")
        for l in covered:
            files = ", ".join(l.test_files[:2])
            if len(l.test_files) > 2:
                files += f" +{len(l.test_files) - 2}"
            lines.append(f"| **{l.requirement_id}** | {l.requirement_title[:50]} | {files} |")
        lines.append("")

    if uncovered:
        lines += [
            "## Recommended Actions",
            "",
            "- Add `# aibrd: <REQ-ID>` comments in test files to improve traceability.",
            "- Run `aibrd generate-tests` to scaffold test cases for uncovered requirements.",
        ]

    return "\n".join(lines)

File: src/core/analyzers/test_linkage.ts
import * as fs from 'fs'
import * as path from 'path'
import { execSync } from 'child_process'

export interface TestLink {
  requirementId: string
  requirementTitle: string
  testFiles: string[]
  testCount: number
  covered: boolean
}

export interface TestLinkageReport {
  coveredCount: number
  uncoveredCount: number
  totalTests: number
  links: TestLink[]
}

/** File extensions considered "test files" */
const TEST_PATTERNS = [
  /\.test\.(ts|tsx|js|jsx|py|java)$/,
  /\.spec\.(ts|tsx|js|jsx|py)$/,
  /Test\.(java|py)$/,
  /_test\.(py|go)$/,
]

function isTestFile(filePath: string): boolean {
  return TEST_PATTERNS.some(p => p.test(filePath))
}

/**
 * Scan the workspace for test files that reference each requirement ID.
 * Looks for inline `// aibrd: BF-001` comments OR the ID appearing in any string.
 */
export function linkTestFiles(
  aibrdDir: string,
  workspaceRoot: string,
  slug?: string
): TestLink[] {
  const ctxPath = slug
    ? path.join(aibrdDir, 'modules', slug, 'CONTEXT.md')
    : path.join(aibrdDir, 'CONTEXT.md')

  if (!fs.existsSync(ctxPath)) return []

  const text = fs.readFileSync(ctxPath, 'utf-8')
  const links: TestLink[] = []

  // Collect all test files in workspace
  let allTestFiles: string[] = []
  try {
    const output = execSync(
      `git -C "${workspaceRoot}" ls-files`,
      { encoding: 'utf-8', maxBuffer: 10 * 1024 * 1024 }
    )
    allTestFiles = output.split('\n').filter(f => f && isTestFile(f))
  } catch {
    // fallback: walk file system (capped at 2000 files)
    const walk = (dir: string, results: string[], depth = 0) => {
      if (depth > 5) return
      for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
        const fullPath = path.join(dir, entry.name)
        if (entry.isDirectory() && !entry.name.startsWith('.') && entry.name !== 'node_modules') {
          walk(fullPath, results, depth + 1)
        } else if (entry.isFile() && isTestFile(fullPath)) {
          results.push(path.relative(workspaceRoot, fullPath))
          if (results.length >= 2000) return
        }
      }
    }
    walk(workspaceRoot, allTestFiles)
  }

  // For each requirement, find which test files mention it
  const reqRegex = /###\s+((?:[A-Z]+-)?(?:BF|AC|TC)-\d+):\s*(.+)/g
  for (const m of text.matchAll(reqRegex)) {
    const id = m[1], title = m[2].trim()
    const matched: string[] = []

    for (const testFile of allTestFiles) {
      const fullPath = path.join(workspaceRoot, testFile)
      try {
        const src = fs.readFileSync(fullPath, 'utf-8')
        if (src.includes(id)) matched.push(testFile)
      } catch { /* skip unreadable */ }
    }

    links.push({
      requirementId: id,
      requirementTitle: title,
      testFiles: matched,
      testCount: matched.length,
      covered: matched.length > 0,
    })
  }

  return links
}

export function buildTestLinkageReport(links: TestLink[]): TestLinkageReport {
  return {
    coveredCount: links.filter(l => l.covered).length,
    uncoveredCount: links.filter(l => !l.covered).length,
    totalTests: links.reduce((s, l) => s + l.testCount, 0),
    links,
  }
}

export function formatTestLinkageReport(report: TestLinkageReport, slug?: string): string {
  const today = new Date().toISOString().split('T')[0]
  const scope = slug ? ` — ${slug}` : ''
  const pct = report.links.length
    ? Math.round((report.coveredCount / report.links.length) * 100)
    : 0

  const lines = [
    `# Test Linkage Report${scope}`,
    `_Generated: ${today}_`,
    '',
    `> **${pct}% requirements covered** · ${report.coveredCount} covered · ${report.uncoveredCount} uncovered · ${report.totalTests} test file references`,
    '',
  ]

  const covered = report.links.filter(l => l.covered)
  const uncovered = report.links.filter(l => !l.covered)

  if (uncovered.length) {
    lines.push('## ❌ Not Covered by Tests', '')
    lines.push('| Requirement | Title |')
    lines.push('|---|---|')
    for (const l of uncovered) {
      lines.push(`| **${l.requirementId}** | ${l.requirementTitle.slice(0, 60)} |`)
    }
    lines.push('')
  }

  if (covered.length) {
    lines.push('## ✅ Covered', '')
    lines.push('| Requirement | Title | Test Files |')
    lines.push('|---|---|---|')
    for (const l of covered) {
      const files = l.testFiles.slice(0, 2).join(', ') + (l.testFiles.length > 2 ? ` +${l.testFiles.length - 2}` : '')
      lines.push(`| **${l.requirementId}** | ${l.requirementTitle.slice(0, 50)} | ${files} |`)
    }
    lines.push('')
  }

  if (uncovered.length) {
    lines.push('## Recommended Actions', '')
    lines.push('- Add `// aibrd: <REQ-ID>` comments in test files to improve traceability.')
    lines.push('- Create test cases for uncovered requirements using `aibrd: Generate Tests`.')
  }

  return lines.join('\n')
}
